Broken in TypeScript < 4.7

## Environment

<!--
All necessary environmental info that will help triage this
-->

<!--
Especially include:
-->

- Linaria version: 
  ```
    "@linaria/babel-preset": "4.4.3",
    "@linaria/core": "4.2.8",
    "@linaria/react": "4.3.6",
    "@linaria/shaker": "4.2.9",
```
- Bundler (+ version): webpack 5
- Node.js version: 18
- OS: ubuntu

## Description

<!--
Describe your issue in detail.
Include screenshots if needed.
If this is a regression, let us know.
-->

```bash
node_modules/@linaria/tags/types/utils/validateParams.d.ts:5:99 - error TS1005: '?' expected.

5 declare type GetParamByName<T> = T extends '*' ? Param : T extends (infer TNames extends ParamName)[] ? Extract<Param, readonly [TNames, ...unknown[]]> : T extends ParamName ? Extract<Param, readonly [T, ...unknown[]]> : never;
                                                                                                    ~

node_modules/@linaria/tags/types/utils/validateParams.d.ts:6:161 - error TS1005: '?' expected.

6 export declare type MapParams<TNames extends ParamConstraints, TRes extends Param[] = []> = TNames extends [infer THead, ...infer TTail extends ParamConstraints] ? THead extends '...' ? [...TRes, ...Params] : MapParams<TTail, [...TRes, GetParamByName<THead>]> : TRes;
                                                                                                                                                                  
```

This is because validateParams uses a 4.7 feature [extends Constraints on infer Type Variables](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-7.html#extends-constraints-on-infer-type-variables)

Without this, only 4.1 is really needed (at least for parsing) as it introduced string template literals.

## Reproducible Demo

<!--
  Let us know how to reproduce the issue.
  Include a code sample or share a project that reproduces the issue.
  Please include you tooling setup eg. webpack config for webpack
  Please follow the guidelines for providing a minimal example: https://stackoverflow.com/help/mcve.
-->

## Solution

You can add an [entry for types less than 4.7 ](https://www.typescriptlang.org/docs/handbook/declaration-files/publishing.html#version-selection-with-typesversions)

```json
  "typesVersions": {
    ">=4.7": {
      "": [
        "lib/index.d.ts"
      ],
      "*": [
        "lib/index.d.ts"
      ]
    },
    ">=4.1": {
      "": [
        "ts4.1/index.d.ts"
      ],
      "*": [
        "ts4.1/index.d.ts"
      ]
    }
  },
```

--- a/.changeset/clean-berries-decide.md
+++ b/.changeset/clean-berries-decide.md
@@ -1,0 +1,5 @@
+---
+'@linaria/tags': patch
+---
+
+Fix TypeScript < 4.7 compatibility (fixes #1227)

--- a/packages/tags/src/utils/validateParams.ts
+++ b/packages/tags/src/utils/validateParams.ts
@@ -7,22 +7,34 @@ export type ParamConstraints =
   | [...ParamConstraint[]]
   | [...ParamConstraint[], '...'];
 
-type GetParamByName<T> = T extends '*'
-  ? Param
-  : T extends (infer TNames extends ParamName)[]
-  ? Extract<Param, readonly [TNames, ...unknown[]]>
-  : T extends ParamName
-  ? Extract<Param, readonly [T, ...unknown[]]>
-  : never;
+// ParamMapping maps each ParamName to its corresponding Param type.
+type ParamMapping = {
+  [K in ParamName]: Extract<Param, readonly [K, ...unknown[]]>; // For each ParamName K, extract the corresponding Param type.
+};
 
+// GetParamByName returns the Param type based on the input type T.
+type GetParamByName<T> = T extends '*'
+  ? Param // If T is '*', return Param type.
+  : T extends keyof ParamMapping // If T is a key in ParamMapping (i.e., a ParamName).
+  ? ParamMapping[T] // Return the corresponding Param type from ParamMapping.
+  : T extends Array<infer TNames> // If T is an array of names.
+  ? TNames extends ParamName // If TNames is a ParamName.
+    ? Extract<Param, readonly [TNames, ...unknown[]]> // Return the corresponding Param type.
+    : never // If TNames is not a ParamName, return never.
+  : never; // If T is none of the above, return never.
+
+// MapParams iteratively maps the input ParamConstraints to their corresponding Param types.
 export type MapParams<
   TNames extends ParamConstraints,
   TRes extends Param[] = []
-> = TNames extends [infer THead, ...infer TTail extends ParamConstraints]
-  ? THead extends '...'
-    ? [...TRes, ...Params]
-    : MapParams<TTail, [...TRes, GetParamByName<THead>]>
-  : TRes;
+> = TNames extends [infer THead, ...infer TTail] // If TNames is a non-empty tuple.
+  ? THead extends '...' // If the first element in the tuple is '...'.
+    ? [...TRes, ...Params] // Append all Params to the result tuple.
+    : MapParams<
+        Extract<TTail, ParamConstraints>, // Extract the remaining ParamConstraints.
+        [...TRes, GetParamByName<Extract<THead, ParamName | '*' | ParamName[]>>] // Append the mapped Param to the result tuple and recurse.
+      >
+  : TRes; // If TNames is an empty tuple, return the result tuple.
 
 export function isValidParams<T extends ParamConstraints>(
   params: Params,

